File: Java/src/GenerateParentheses.java
import java.util.*;

/**
 * Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
 * <p>
 * For example, given n = 3, a solution set is:
 * <p>
 * [
 * "((()))",
 * "(()())",
 * "(())()",
 * "()(())",
 * "()()()"
 * ]
 * <p>
 * Accepted.
 */
public class GenerateParentheses {

    public List<String> generateParenthesis(int n) {
        Set<String> set = new HashSet<>();
        List<String> list = new ArrayList<>();

        if (n <= 0) {
            return list;
        }
        if (n == 1) {
            list.add("()");
            return list;
        }

        for (String s : generateParenthesis(n - 1)) {
            StringBuilder sb = new StringBuilder();
            char[] chars = s.toCharArray();

            for (int j = 0; j < chars.length; j++) {
                sb.append(chars[j]);
                if (chars[j] == '(') {
                    StringBuilder builder = new StringBuilder(sb);
                    builder.append("()");
                    for (int k = j + 1; k < chars.length; k++) {
                        builder.append(chars[k]);
                    }
                    set.add(builder.toString());
                }
            }

            set.add(s + "()");
        }

        list.addAll(set);
        return list;
    }

}

File: Kotlin/src/GenerateParentheses.kt
/**
 * Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
 *
 * For example, given n = 3, a solution set is:
 *
 * [
 * "((()))",
 * "(()())",
 * "(())()",
 * "()(())",
 * "()()()"
 * ]
 *
 * Accepted.
 */
class GenerateParentheses {

    fun generateParenthesis(n: Int): List<String> {
        val set = mutableSetOf<String>()
        val list = mutableListOf<String>()

        if (n <= 0) {
            return list
        }
        if (n == 1) {
            return list.apply {
                add("()")
            }
        }

        generateParenthesis(n - 1).forEach { s ->
            val sb = StringBuilder()
            val chars = s.toCharArray()

            for (j in chars.indices) {
                sb.append(chars[j])
                if (chars[j] == '(') {
                    val builder = StringBuilder(sb)
                    builder.append("()")
                    for (k in j + 1 until chars.size) {
                        builder.append(chars[k])
                    }
                    set.add(builder.toString())
                }
            }

            set.add(s + "()")
        }

        return list.apply {
            addAll(set)
        }
    }

}
